### PR TITLE
Rewrite Fiber's type system, with a few improvements

### PR DESCRIPTION
**Pull request by Pyrofab**
This PR is based on the following design document: https://discordapp.com/channels/691048196507697302/698236292147970118/701881290458267661

*TL;DR : custom types do not belong to the intermediate representation and must therefore be confined to helper interfaces.*

The main change compared to the proposed design (besides the `ConfigType` -> `SerializableType` rename) is the replacement of `TypeConverter` with a hierarchy of `ConfigType` mirroring the hierarchy of `SerializableType`. I figured that if we wanted a smooth API, we should make all the information available from a single type, rather than split between `TypeConverter` and `SerializableType`.

**New class diagram :**

DataFiberUpper_new_classes

This PR is a bit audacious, but I believe it should be helpful in making Fiber more stable and generally helpful.

### CODE REVIEW

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/EnumSerializableType.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package me.zeroeightsix.fiber.api.schema.type;
+
+import me.zeroeightsix.fiber.api.serialization.TypeSerializer;
+import me.zeroeightsix.fiber.impl.constraint.EnumConstraintChecker;
+
+import java.util.*;
+
+public final class EnumSerializableType extends SerializableType<String> {
+
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/EnumSerializableType.java:9]: This can work with Java enums, no?
`class EnumSerializableType<E extends Enum<E>> extends SerializableType<E>`

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/EnumSerializableType.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    private final Set<String> validValues;
+
+    public EnumSerializableType(String... validValues) {
+        this(new HashSet<>(Arrays.asList(validValues)));
+    }
+
+    public EnumSerializableType(Set<String> validValues) {
+        super(String.class, EnumConstraintChecker.instance());
+        this.validValues = Collections.unmodifiableSet(validValues);
+    }
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/EnumSerializableType.java:19]: `Collections.unmodifiableSet` does not guarantee immutability.

File: src/main/java/me/zeroeightsix/fiber/impl/constraint/ConstraintChecker.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	 * if the input value is invalid.
+	 *
+	 * @param cfg the type configuration to test against
+	 * @param value the value
+	 * @return {@code true} if {@code value} satisfies the constraint
+	 */
+	public abstract TypeCheckResult<V> test(T cfg, V value);
+
+	public abstract boolean comprehends(T cfg, T cfg2);
+
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/impl/constraint/ConstraintChecker.java:31]: What is this used for?

File: src/main/java/me/zeroeightsix/fiber/api/builder/ConfigTreeBuilder.java [hunk 4]
```diff
@@ -25,20 +27,10 @@
...
+ *         .withValue("A", ConfigTypes.INTEGER, 10)
  *         .fork("child")
- *             .beginValue("A", 10)
- *                 .beginConstraints()
- *                     .composite(CompositeType.OR)
- *                         .atLeast(3)
- *                         .atMost(0)
- *                     .finishComposite()
- *                 .finishConstraints()
+ *             .beginValue("drops", ConfigTypes.STRING_SET, new HashSet<>(Arrays.asList("diamond", "cactus"))
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/api/builder/ConfigTreeBuilder.java:32]: `ConfigTypes.STRING_SET` does not exist.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/RecordConfigType.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+import me.zeroeightsix.fiber.api.annotation.processor.ConstraintAnnotationProcessor;
+import me.zeroeightsix.fiber.api.schema.type.RecordSerializableType;
+import me.zeroeightsix.fiber.api.tree.ConfigBranch;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.util.function.Function;
+
+public final class RecordConfigType<R, E> extends ConfigType<R, ConfigBranch, RecordSerializableType> {
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/RecordConfigType.java:11]: `E` seems to be unused.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/TypeCheckResult.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    private final boolean passed;
+    @Nullable
+    private final V correctedValue;
+
+    public TypeCheckResult(boolean passed, @Nullable V correctedValue) {
+        this.passed = passed;
+        this.correctedValue = correctedValue;
+    }
+
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/TypeCheckResult.java:35]: This should be made `private`.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                    }
+                    return Collections.unmodifiableList(ret);
+                }
+        );
+    }
+
+    public static <E0, E, U extends ConfigType<E, E0, ?>> ListConfigType<Set<E>, E0> makeSet(U elementType) {
+        return makeList(elementType).withUniqueElements().derive(Set.class, HashSet::new, ArrayList::new);
+    }
+
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java:88]: This can be changed not to go through an intermediate `List`.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    @SuppressWarnings("unchecked")
+    public static <E0, E, U extends ConfigType<E, E0, ? extends SerializableType<E0>>> ListConfigType<E[], E0> makeArray(U elementType) {
+        E[] z = (E[]) Array.newInstance(elementType.getRuntimeType(), 0);
+        Class<E[]> arrType = (Class<E[]>) z.getClass();
+        // account for arrays of primitives - we must avoid autocasting to Object[]
+        return (ListConfigType<E[], E0>) (ListConfigType<?, ?>) makeList((ConfigType<Object, Object, ?>) elementType).derive((Class<? super Object>) (Class<?>) arrType, l -> TypeMagic.unbox(l.toArray((Object[]) z)), a -> Arrays.asList((E[]) TypeMagic.box(a)));
+    }
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java:95]: What about `ListConfigType<int[], Integer>`? If you want primitives, you may have to write a specialized method for each primitive type.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/SerializableType.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+ * @param <T> the actual platform type used to store the serializable data
+ * @see ConfigType
+ * @see BooleanSerializableType
+ * @see EnumSerializableType
+ * @see ListSerializableType
+ * @see MapSerializableType
+ * @see DecimalSerializableType
+ * @see RecordSerializableType
+ * @see StringSerializableType
+ */
```
[the remark this answers is not in the document]
Comment (Pyrofab) [reply src/main/java/me/zeroeightsix/fiber/api/schema/type/SerializableType.java:29]: Counterpoint: IDEs and other text editors don't show the "known subclasses" section.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/EnumSerializableType.java (hunk 1 above)
Comment (Pyrofab) [reply src/main/java/me/zeroeightsix/fiber/api/schema/type/EnumSerializableType.java:9]: I don't like that very much as it means you get custom types in the intermediate representation. A `SerializableType` should be deserializable with no context from the application that created it.

File: src/main/java/me/zeroeightsix/fiber/api/builder/ConfigTreeBuilder.java (hunk 4 above)
Comment (Pyrofab) [reply src/main/java/me/zeroeightsix/fiber/api/builder/ConfigTreeBuilder.java:32]: Good catch.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/RecordConfigType.java (hunk 5 above)
Comment (Pyrofab) [reply src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/RecordConfigType.java:11]: Looks like it.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java (hunk 8 above)
Comment (Pyrofab) [reply src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java:95]: This was absolutely ridiculous, ditched it for something vaguely reasonable.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/EnumSerializableType.java (hunk 1 above)
Comment (kvverti) [reply src/main/java/me/zeroeightsix/fiber/api/schema/type/EnumSerializableType.java:9]: Fair enough. This way configs are completely defined by their schemas.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/SerializableType.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kvverti) [reply src/main/java/me/zeroeightsix/fiber/api/schema/type/SerializableType.java:29]: Useful for sealed hierarchies, I see.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        );
+    }
+
+    public static <E0, E, U extends ConfigType<E, E0, ? extends SerializableType<E0>>> ListConfigType<E[], E0> makeArray(U elementType) {
+        @SuppressWarnings("unchecked") Class<E[]> arrType = (Class<E[]>) Array.newInstance(elementType.getRuntimeType(), 0).getClass();
+        return new ListConfigType<>(
+                new ListSerializableType<>(elementType.getSerializedType()),
+                arrType,
+                l -> {
+                    @SuppressWarnings("unchecked") E[] ret = (E[]) Array.newInstance(elementType.getRuntimeType(), l.size());
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java:111]: Specifically this cast may fail when `elemetType.getRuntimeType()` represents a primitive type.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+                    for (int i = 0; i < l.size(); i++) {
+                        ret[i] = elementType.toRuntimeType(l.get(i));
+                    }
+                    return ret;
+                },
+                arr -> {
```
Comment (kvverti) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java:117]: There is an implicit cast here that may also fail when the runtyoe type represents an array of primitive.

**Comment by kvverti** [comment]
I'm looking into primitive specializations now.

File: src/main/java/me/zeroeightsix/fiber/api/builder/ConfigLeafBuilder.java [hunk 12]
```diff
@@ -1,62 +1,69 @@
...
+
+    public static <T> ConfigLeafBuilder<T, T> create(ConfigTreeBuilder parentNode, @Nonnull String name, @Nonnull SerializableType<T> type) {
+        return new ConfigLeafBuilder<>(parentNode, name, type, Function.identity(), Function.identity());
+    }
 
     @Nonnull
-    protected final Class<T> type;
+    protected final SerializableType<T> type;
+    protected final Function<T, R> f;
+    protected final Function<R, T> f0;
```
Comment (zeroeightysix) [inline src/main/java/me/zeroeightsix/fiber/api/builder/ConfigLeafBuilder.java:42]: These need proper names as well.

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/StringSerializableType.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        return this.minLength;
+    }
+
+    /**
+     * Specifies a maximum string length.
+     *
+     * <p> Values must be of equal or shorter length than the returned value to satisfy the constraint.
+     * For example: if the max length is 3.
+     * <ul>
+     *     <li> {@code "AB"} and {@code "ABC} would satisfy the constraint</li>
```
Comment (zeroeightysix) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/StringSerializableType.java:46]: `"ABC"` instead of `"ABC`

File: src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+                    List<E0> ret = new ArrayList<>();
+                    for (E e : l) {
+                        ret.add(elementType.toPlatformType(e));
+                    }
+                    return Collections.unmodifiableList(ret);
+                }
+        );
+    }
+
+    public static <E0, E, U extends ConfigType<E, E0, ?>> ListConfigType<Set<E>, E0> makeSet(U elementType) {
```
Comment (zeroeightysix) [inline src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java:85]: `E0` is not a very pretty generic name, maybe use `R` and `S` (from `ConfigType`) instead?

File: src/main/java/me/zeroeightsix/fiber/api/tree/HasValue.java [hunk 15]
```diff
@@ -12,10 +12,5 @@
      */
     T getValue();
 
-    /**
-     * Returns the class of the type of the value being held
-     * @return the class of the type of the value
-     */
     Class<T> getType();
```
Comment (zeroeightysix) [inline src/main/java/me/zeroeightsix/fiber/api/tree/HasValue.java:15]: Why remove this doc?

File: src/test/java/me/zeroeightsix/fiber/api/NodeOperationsTest.java [hunk 16]
```diff
@@ -18,51 +20,51 @@
...
-                .withValue("A", Integer.class, 10)
+                .withValue("A", ConfigTypes.INTEGER, 10)
                 .build();
 
         ConfigTreeBuilder nodeTwo = ConfigTree.builder();
 
         NodeOperations.moveChildren(treeOne, nodeTwo);
 
-        testNodeFor(nodeTwo, "A", Integer.class, 10);
+        testNodeFor(nodeTwo, "A", ConfigTypes.INTEGER.getSerializedType(), BigDecimal.valueOf(10));
```
Comment (zeroeightysix) [inline src/test/java/me/zeroeightsix/fiber/api/NodeOperationsTest.java:30]: Maybe use `BigDecimal.TEN` here as well (sorry for the nitpick)

File: src/test/java/me/zeroeightsix/fiber/api/builder/constraint/ConstraintsBuilderTest.java [hunk 17]
```diff
@@ -1,24 +1,7 @@
...
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
-import java.util.function.Predicate;
-
-import static org.junit.jupiter.api.Assertions.*;
-
 class ConstraintsBuilderTest {
+/* TODO convert to a ConfigTypeTest
```
Comment (zeroeightysix) [inline src/test/java/me/zeroeightsix/fiber/api/builder/constraint/ConstraintsBuilderTest.java:4]: Having an entire class commented out in `master` would be a bit ugly, mind either removing the class or performing the conversion?

File: src/main/java/me/zeroeightsix/fiber/api/tree/HasValue.java (hunk 15 above)
Comment (Pyrofab) [reply src/main/java/me/zeroeightsix/fiber/api/tree/HasValue.java:15]: Mistake, I removed this method then added it back

File: src/test/java/me/zeroeightsix/fiber/api/builder/constraint/ConstraintsBuilderTest.java (hunk 17 above)
Comment (Pyrofab) [reply src/test/java/me/zeroeightsix/fiber/api/builder/constraint/ConstraintsBuilderTest.java:4]: Looks like I had forgotten about that, will do the conversion

File: src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java [hunk 18]
```diff
@@ -1,22 +1,84 @@
...
-import org.junit.jupiter.api.Test;
+import static java.util.stream.Collectors.toList;
+import static org.junit.jupiter.api.Assertions.*;
+
+class ConfigTypesTest {
+
+	@DisplayName("Test numerical constraints")
+	@Test
+	void testNumericalConstraints() {
+		assertThrows(NullPointerException.class, () -> ConfigTypes.UNBOUNDED_DECIMAL.withIncrement(5), "Invalid increment");
```
Comment (kvverti) [inline src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java:27]: Throws `NullPointerException`?

File: src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java [hunk 19]
```diff
@@ -1,22 +1,84 @@
...
+		assertFalse(type.accepts(BigDecimal.valueOf(0.5)));
+	}
+
+	@DisplayName("Test array type constraints")
+	@Test
+	void testArrayConstraints() {
+		DecimalSerializableType elementType = ConfigTypes.INTEGER.withValidRange(3, 10, 2).getSerializedType();
+		ListSerializableType<BigDecimal> type = new ListSerializableType<>(elementType, 0, 3, false);
+		Predicate<Integer[]> config = arr -> type.accepts(Arrays.stream(arr).map(BigDecimal::new).collect(Collectors.toList()));
+
```
Comment (kvverti) [inline src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java:52]: `BigDecimal::valueOf` and plain `toList()`

Comment (Pyrofab) [reply src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java:52]: plain `toList()` ?

File: src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java (hunk 18 above)
Comment (kvverti) [reply src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java:27]: `IllegalStateException` would make more sense here.

File: src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java (hunk 19 above)
Comment (kvverti) [reply src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java:52]: `toList()` is already statically imported.

Comment (Pyrofab) [reply src/test/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypesTest.java]: Well, since we are mostly avoiding static imports in this project I'll switch it to a regular import. We can always decide on a different convention later.
